### DISCUSSION TITLE
Counters for multiline row/cells

### DISCUSSION THREAD

**Discussion by wwojciech** [lvjr/tabularray#485] [Ideas]
It seems that standard counters do not work as expected in multiline setup, i.e. this code 

```
\begin{tblr}{hlines, vlines}
    \SetCell[r=2]{c} {\arabic{rownum}} & A \\ & B \\
    \SetCell[r=2]{c} {\arabic{rownum}} & C \\ & D
\end{tblr}
```
gives:
<img width="109" alt="a" src="https://github.com/lvjr/tabularray/assets/11532997/05281240-159c-4f76-ba19-d11dd1e00851">

while I would rather expect:
<img width="109" alt="b" src="https://github.com/lvjr/tabularray/assets/11532997/2c8dc21f-c80e-4fb5-ad04-4d6621e9fd7a">

As a workaround, I can use `\UseTblrLibrary{counter}`, however, the counter is then not reset across different tables (I would probably need to reset it every time, before a new table?). 

Eventually, I could use the construct like below, but it does not work:
```
\begin{tblr}{hlines, vlines}
    \SetCell[r=2]{c} {\arabic{rownum}} & A \\ & B \\
    \SetCell[r=2]{c} {\arabic{rownum+1}} & C \\ & D
\end{tblr}
```

**Comment by muzimuzhi** [comment]
Sounds like a request for "real row/column number" counters.

`\AddToHook{env/<envname>/begin}{<code>}` can be used to add code to the `\begin{<envname>}` of an environment. It's documented in `texdoc lthooks-doc`.

BTW you can use `\inteval{\value{rownum}+1}` to do integer calculations.

`\inteval` is added to the kernel since LaTeX2e 2022-06-01, see `texdoc ltnews`, issue 35. `\inteval{<int expr>}` is let to `\int_eval:n` expl3 function and the latter is a thin wrapper around `\the\intexpr <int expr>\relax`, thus the supported syntax is limited, see `texdoc etex` and this [TeX-SX answer](https://tex.stackexchange.com/a/713246).

**Reply by wwojciech** [reply]
So, I see that `\inteval{\value{rownum}/2}`rounds the result up, i.e. for instance for `rownum=3` it returns `2`. Do you know if there is any explicit way or ceiling?

**Reply by muzimuzhi** [reply]
If you meant floor (round towards negative infinity), `\inteval{(\value{rownum}-1)/2}` does the trick.

More powerful thus slower `\fpeval{floor(\value(rownum), 0)}` gives the same result. `floor` floating-point math function is documented in `texdoc interface3`, sec. 29.12.3 "Operations".

**Reply by wwojciech** [reply]
Thanks, I meant ceiling (the smallest integer not less than a given number, or rounding towards positive infinity, if you prefer).
Actually `\inteval{\value{rownum}/2}` gives expected results, i.e. 1, 2, ... for "real" row numbers. However, I think, syntacticly, it would be better to use `\fpeval{ceil(\value{rownum}/2)}`.

**Comment by muzimuzhi** [comment]
I filed a feature request, see
- #486.

**Reply by wwojciech** [reply]
Thanks lot for your valuable response and for the feature request!
